From: Yuan Tao <ty@wevs.org>
Date: Sun, 9 Aug 2020 09:02:55 +0800
Subject: rockchip:nanopi-r2s-improve-boot-failed

When booting on some SD cards an error message appears as:
"spl: mmc init failed with error: -95"

Solutions:
Add startup-delay-us to vcc_sd and vcc_sdio.
Disable the jtag switching on the RK3328.
This will improve the problem that in some SD cards are failing to boot up.

Tested environment:
SD Card: Netac Extreme P500 Pro 32GB

Signed-off-by: Yuan Tao <ty@wevs.org>
---

diff --git a/arch/arm/dts/rk3328-nanopi-r2s.dts b/arch/arm/dts/rk3328-nanopi-r2s.dts
--- a/arch/arm/dts/rk3328-nanopi-r2s.dts
+++ b/arch/arm/dts/rk3328-nanopi-r2s.dts
@@ -32,6 +32,7 @@
 		regulator-name = "vcc_sd";
 		regulator-min-microvolt = <3300000>;
 		regulator-max-microvolt = <3300000>;
+		startup-delay-us = <100000>;
 		vin-supply = <&vcc_io>;
 	};
 
@@ -49,6 +50,7 @@
 		regulator-name = "vcc_sdio";
 		regulator-settling-time-us = <5000>;
 		regulator-type = "voltage";
+		startup-delay-us = <2000>;
 		vin-supply = <&vcc_io>;
 	};
 

diff --git a/arch/arm/mach-rockchip/rk3328/rk3328.c b/arch/arm/mach-rockchip/rk3328/rk3328.c
--- a/arch/arm/mach-rockchip/rk3328/rk3328.c
+++ b/arch/arm/mach-rockchip/rk3328/rk3328.c
@@ -50,9 +50,14 @@ int arch_cpu_init(void)
 {
 #ifdef CONFIG_SPL_BUILD
 	/* We do some SoC one time setting here. */
+	struct rk3328_grf_regs * const grf = (void *)GRF_BASE;
 
 	/* Disable the ddr secure region setting to make it non-secure */
 	rk_setreg(FW_DDR_CON_REG, 0x200);
+
+	/* Enable force to sdmmc0, jtag_tclk/tms iomuxed with sdmmc0_d2/d3 */
+	rk_clrreg(&grf->soc_con[4], 1 << 12);
+
 #endif
 	return 0;
 }
 
